TODO/Android-Studio-Tips.md

```diff
@@ -1,832 +1,542 @@
 
 ![](https://lh3.googleusercontent.com/-bP5WOVMfp7A/U_cpQi0bvhI/AAAAAAAAM9c/dcvvJu1US40/w265-h103-no/10-deleteline.gif)
 
+# 13. 断点失效（Disable Breakpoints）
 
-# Disable Breakpoints
-
-This will disable the breakpoint. Particularly useful when you have some complicated conditional or logging breakpoint that you don't need right now but don't want to recreate next time.
-
-Shortcut: 
-  - Mouse: **Alt+LeftClick** on an existing breakpoint in the left gutter
-  - There is no keyboard shortcut but you can create one if you use it often enough.﻿
+- 这个操作将使得断点失效。当处在一种复杂的状态或者某个断点你现在不需要，但是下次又不用重新创建，该操作是很方便的。
+- **调用：**按住Alt，然后单击断点即可。
 
 ![](https://lh3.googleusercontent.com/-hNk0kuL1WBM/VBbQXamG8-I/AAAAAAAANeM/ynfSJ5hqCvA/w365-h235-no/25-diablebreakpoint.gif)
 
+# 14. 行复制（Duplicate Line）
 
-# Duplicate Line
-
-Just as it says: it will copy the current line and paste it below without interfering with your clipboard. It can be pretty useful when it's used with the move line shortcut!
-
-Shortcut: **cmd+d(mac)** or **ctrl+d(windows/linux)﻿**
+- **描述：**复制当前行，并粘贴到下一行。
+- **快捷键：**Cmd + D(OS X)、Ctrl + D(Windows/Linux)
 
 ![](https://lh6.googleusercontent.com/-1dno1jn2Pcg/U_sfhOxXTkI/AAAAAAAANC8/8sl3TVz1dAo/w265-h103-no/11-duplicate_lines.gif)
 
+# 15. 编写正则表达式（Edit Regex）
 
-# Edit Regex
+- **描述：**使用Java编写正则表达式是一件很困难的事，主要原因是：
+    - 你必须得避开反斜杠；
+    - 说实话，正则很难；
+    - 看第二条。
 
-Editing a regex in Java is a bit difficult for a couple of reasons
-1) You have to escape the backslashes
-2) Let's face it, regex are hard!
-3) See point #2
-
-So what can the IDE do to help us? Give us a nice interface to code and test regexes of course!
-
-Shortcut: **Alt+Enter → check regexp**
+IDE能帮我们干点啥呢？当然是一个舒服的界面来编写和测试正则啦~
+- **快捷键：**Alt + Enter → check regexp
 
 ![](https://lh4.googleusercontent.com/-zinVQioQi0c/VGX3txYe0iI/AAAAAAAAO5c/D5nhpSSyImk/w419-h170-no/68-checkregexp.gif)
 
+# 16. 使用Enter和Tab进行代码补全的差别（Enter vs Tab for Code Completion）
 
-# Enter vs Tab for Code Completion
-
-There is an interesting difference whether you use code completion with tab or with enter. 
-
-Using enter will complete the statement as you would expect.
-Using tab will complete the statement and delete everything forward until the next dot, parentheses, semicolon or space.
-
-Shortcut (during code completion): **enter or tab**
+- **描述：**代码补全时，可以使用Enter或Tab来进行补全操作，但是两者是有差别的。
+- 使用Enter时：从光标处插入补全的代码，对原来的代码不做任何操作。
+- 使用Tab时：从光标处插入补全的代码，并删除后面的代码，直到遇到点号、圆括号、分号或空格为止。
 
 ![](https://lh3.googleusercontent.com/-zkDYRijGp4A/VD0KtdkrqFI/AAAAAAAAOJE/wEr134jmFxE/w252-h123-no/45-codecompletionentertab.gif)
 
+# 17. 计算表达式（Evaluate Expression）
 
-# Evaluate Expression
-
-This is used to inspect a variable's content and evaluate pretty much any valid java expression right there, right now.
-Just be aware that if you are mutating state, it will stay that way when you resume the execution of the program.
-
-Shortcut: **Alt+F8﻿**
-
+- **描述：**这是一个很有用的操作用来审查某个变量的内容和计算表达式的值。
```

Comment:
这个操作可以用来查看变量的内容并且计算几乎任何有效的java表达式。

Just be aware that if you are mutating state, it will stay that way when you resume the execution of the program.
需要注意的是，如果你修改了变量的状态，这个状态在你恢复代码执行后依然会保留。